library/core/src/primitive_docs.rs

```diff
@@ -268,6 +268,47 @@ mod prim_bool {}
 /// [`Debug`]: fmt::Debug
 /// [`default()`]: Default::default
 ///
+/// # Never type fallback
+///
+/// When the compiler sees a value of type `!` it implicitly inserts a coercion (if possible),
+/// to allow type check to infer any type:
+///
+/// ```rust,ignore (illustrative-and-has-placeholders)
+/// // this
+/// let x: u8 = panic!();
+///
+/// // is (essentially) turned by the compiler into
+/// let x: u8 = absurd(panic!());
+///
+/// // where absurd is a function with the following signature
+/// // (it's sound, because `!` always marks unreachable code):
+/// fn absurd<T>(_: !) -> T { ... }
+// FIXME: use `core::convert::absurd` here instead, once it's merged
+/// ```
+///
+/// While it's convenient to be able to use non-diverging code in one of the branches (like
+/// `if a { b } else { return }`) this could lead to compilation errors:
+///
+/// ```compile_fail
+/// // this
+/// { panic!() };
+///
+/// // gets turned into this
+/// { absurd(panic!()) }; // error: can't infer the type of `absurd`
+/// ```
+///
+/// To prevent such errors, compiler remembers where it inserted `absurd` calls, and if it can't
+/// infer their type, it sets the type to the fallback type. `{ absurd::<Fallback>(panic!()) };`.
+/// This is what is known as "never type fallback".
```

Comment:
The sudden mention of `Fallback` surprised me a little. I think we can spare a little vertical space to make it clearer that this type alias is designating an arbitrary substitution:
```suggestion
/// To prevent such errors, the compiler remembers where it inserted `absurd` calls, and
/// if it can't infer their type, it sets the type to the fallback type:
/// ```rust, ignore
/// type Fallback = /* An arbitrarily selected type! */;
/// { absurd::<Fallback>(panic!()) }
/// ```
///
/// This is what is known as "never type fallback".
```